sources/engine/Xenko.Rendering/Rendering/Lights/LightSkyboxRenderer.cs

```diff
@@ -90,6 +90,8 @@ private class LightSkyBoxShaderGroup : LightShaderGroup
             private ObjectParameterKey<Texture> specularCubeMapkey;
             private ValueParameterKey<float> specularMipCountKey;
 
+            private Color3[] sphericalColors;
```

Comment:
Is it ok? Could that instance be shared by different components of a renderer or pass, in which case it shouldn't be a field but a local variable. In other words, could `ApplyViewParameters()` be called concurrently with different `RenderDrawContext`?

I am not really knowledgeable in that low level layer that is the rendering. But I know that some parts are multi-threaded. @xen2 any thoughts?

Reply:
Actually that's a good point. Looking through the call chain, this is not placed in a multi-threaded/Dispatcher section of code. Not sure if it's a safe or unsafe assumption to make, though?
Maybe a side note, but these renders also make similar assumptions, with referencing some class field:
https://github.com/stride3d/xenko/blob/c1522239be6e424a3d0362bd726512a3afb69361/sources/engine/Xenko.Rendering/Rendering/Lights/LightShaderGroupDynamic.cs#L132
https://github.com/stride3d/xenko/blob/717771d680686003563671281f4a817b629e7d9d/sources/engine/Xenko.Voxels/Voxels/Light/LightVoxelRenderer.cs#L203
Maybe these should also be considered bugs, unless there's some other subtle point I'm missing.

Is there a better way to try to reduce the array alloc? A jagged array where the first index is the viewIndex?

Reply:
You could use stackalloc

Reply:
As it stands, stackalloc doesn't work because the array is created inside a method and needs to be returned.
Having said that, while trying to find a way to make stackalloc work you've given me the idea that I should probably just pass the other ParameterCollection in a method and do some kind of values transfer which will skip creating an array anyway.

Reply:
It can work, you just have more work to do.
From the calling scope, get the amount of values, stackalloc of that amount, pass the stackalloc pointer and the max amount to the method, that method pushes its values to the pointer and returns the final amount(which in this case should be the same but if the caller used a larger stackalloc/buffer than the amount of values it could be useful).
You could also create a struct enumerator instead of all of this, depending on your cases that might be good enough.

Reply:
Understood. I didn't want to do it this way as determining the size requires searching for the correct underlying key which would happen twice (once for getting the size, then another for the copying).
I have force pushed a version that just passes the other ParameterCollection and copy directly. I guess this way also does not risk potentially blowing the stack (though extremely unlikely).